android/src/main/java/com/bitmovin/player/reactnative/BitmovinBaseModule.kt

```diff
@@ -0,0 +1,106 @@
+        crossinline block: RejectPromiseOnExceptionBlock.() -> R,
+    ) {
+        val uiManager = runAndRejectOnException { uiManager } ?: return
+        uiManager.addUIBlock {
+            resolveOnCurrentThread { block() }
+        }
+    }
+
+    protected val RejectPromiseOnExceptionBlock.playerModule: PlayerModule get() = context.playerModule
+        ?: throw IllegalArgumentException("PlayerModule not found")
+
+    protected val RejectPromiseOnExceptionBlock.uiManager: UIManagerModule get() = context.uiManagerModule
+        ?: throw IllegalStateException("UIManager not found")
+
+    protected val RejectPromiseOnExceptionBlock.sourceModule: SourceModule get() = context.sourceModule
+        ?: throw IllegalStateException("SourceModule not found")
+
+    protected val RejectPromiseOnExceptionBlock.offlineModule: OfflineModule get() = context.offlineModule
+        ?: throw IllegalStateException("OfflineModule not found")
+
+    protected val RejectPromiseOnExceptionBlock.drmModule: DrmModule get() = context.drmModule
+        ?: throw IllegalStateException("DrmModule not found")
+
+    fun RejectPromiseOnExceptionBlock.getPlayer(
+        nativeId: NativeId,
+        playerModule: PlayerModule = this.playerModule,
+    ): Player = playerModule.getPlayerOrNull(nativeId) ?: throw IllegalArgumentException("Invalid PlayerId $nativeId")
+
+    fun RejectPromiseOnExceptionBlock.getSource(
+        nativeId: NativeId,
+        sourceModule: SourceModule = this.sourceModule,
+    ): Source = sourceModule.getSourceOrNull(nativeId) ?: throw IllegalArgumentException("Invalid SourceId $nativeId")
+}
+
+/** Run [block], returning it's return value. If [block] throws, [Promise.reject] [this] and return null. */
+inline fun <T, R> TPromise<T>.runAndRejectOnException(block: RejectPromiseOnExceptionBlock.() -> R): R? = try {
+    RejectPromiseOnExceptionBlock.block()
+} catch (e: Exception) {
+    reject(e)
+    null
+}
+
+/**
+ * [TPromise.resolve] [this] with [block] return value.
+ * If [block] throws, [Promise.reject] [this] with the [Throwable].
+ */
+inline fun <T> TPromise<T>.resolveOnCurrentThread(
+    crossinline block: RejectPromiseOnExceptionBlock.() -> T,
+): Unit = runAndRejectOnException { this@resolveOnCurrentThread.resolve(block()) } ?: Unit
+
+/** Receiver of code that can safely throw when resolving a [Promise]. */
+object RejectPromiseOnExceptionBlock
+
+/** Compile time wrapper for Promises to type check the resolved type [T]. */
+@JvmInline
+value class TPromise<T>(val promise: Promise) {
+    // Promise only support built-in types. Functions that return [Unit] must resolve to `null`.
+    fun resolve(value: T): Unit = promise.resolve(value.takeUnless { it is Unit })
+    fun reject(throwable: Throwable) {
+        Log.e(MODULE_NAME, "Failed to execute Bitmovin method", throwable)
+        promise.reject(throwable)
+    }
+}
+
+inline val Promise.int get() = TPromise<Int>(this)
```

Comment:
I think the idea behind the convenience properties is to also show which types are possible. (the supported types can not be nicely generalized)

+1 for the function instead of the property, but this should not block the PR.

Reply:
> `promise.resolveOnUiThreadAs<Int>`

This was my first version, but rejected due to the following drawback:
 - the type can be invalid (only primitive types, array and map are supported)
 - the type is autodeduced, so the explicit type can be omitted
 - the goal of this is to make it explicit to the reader which type is returned
 - it's not possible to constraint the types at compile time with this method
 - more verbose (eg: `<ReadableMap?>`)
 - constraining the type is more composable

Reply:
> `fun Promise.asIntPromise()`

I kept the naming short because it's used in many places and it explicit and type safe.
I'm not a fan of repeating `promise` twice.
Should we bring it to the JFX? (naming is hard and I want this pr merged ASAP)

Reply:
I think the `inline fun Promise.asIntPromise()` would be more dogmatic, and still on the same level of type safety as the property.
I would though prefer to merge this PR rather sooner than later. Before this starts to be a discussion, let's get this PR merged and iterate on the system in follow up PRs.

Reply:
I'm fine keeping this as is 👍